discovery/README.md

```diff
@@ -0,0 +1,123 @@
+### Service Discovery
+
+This directory contains the service discovery (SD) component of Prometheus.
+
+
+
+## Design of a Prometheus SD
+
+There are many requests to add new SDs to Prometheus, this section looks at
+what makes a good SD and covers some of the common implementation issues.
+
+### Does this make sense as an SD?
+
+The first question to be asked is does it make sense to add this particular
+SD? An SD mechanism should be reasonably well established, and at a minimum in
+use across multiple orgnaisations. It should allow discovering of machines
+and/or services running somewhere. When exactly an SD is popular enough to
+justify being added to Prometheus natively is an open question.
+
+It should not be a brand new SD mechanism, or a variant of an established
+mechanism. We want to integrate Prometheus with the SD that's already there in
+your infrastructure, not invent yet more ways to do service discovery. We also
+do not add mechanisms to work around users lacking service discovery and/or
+configuration management infrastructure.
+
+SDs that merely discover other applications running the same software (e.g.
+talk to one Kafka or Cassandra server to find the others) are not service
+discovery. In that case the SD you should be looking at is whatever decides
+that a machine is going to be a Kafka server, likely a machine database or
+configuration management system.
+
+If something is particlarly custom or unusual, `file_sd` is the generic
+mechanism provided for users to hook in. Generally with Prometheus we offer a
+single generic mechanism for things with infinite variations, rather than
+trying to support everything natively (see also, alertmanager webhook, remote
+read, remote write, node exporter textfile collector). For example anything
+that would involve talking to a relational database should use `file_sd`
+instead.
+
+For configuration management systems like Chef, while they do have a
+database/API that'd in principle make sense to talk to for service discovery,
+the idomatic approach is to use Chef's templating facilities to write out a
+file for use with `file_sd`.
```

Comment:
I don't think user guides belong buried among source code, when we figure out what we're doing with user guides this can go with them though most of what such a user would need is the yaml format.